### PR TITLE
DOF table subset

### PR DESCRIPTION
**Pull request by endJunction**
This is preparation for the Neumann type boundary conditions.

The PR consists of three parts:
 - adding `MeshSubset::setIntersectionByNodes()`,
 - adding `MeshComponentMap::getSubset()` where the nodes subset is supplied by the set intersection by nodes function,
 - adding `LocalToGlobalIndexMap::deriveBoundaryConstrainedMap()` which is using the get subset function for mesh component map.

This feature in the `LocalToGlobalIndexMap` is used by the Neumann-type boundary condition implementation: The integration of the Neumann-type boundary conditions is essentially another process (like the groundwater flow process) contributing to the global matrix and vectors but it is executed only for boundary elements, for which a smaller DOF table is needed.

You can see ongoing development of the Neumann-type boundary conditions on [endJunction/NeumannBC](https://github.com/endJunction/ogs/tree/NeumannBC) branch.

WIP test for `LocalToGlobalIndexMap::deriveBoundaryConstrainedMap()` missing.

### CODE REVIEW

**Comment by wenqing** [comment]
@endJunction We changed the same file. Do you want to merge the stuff of Neumann BC first?

**Comment by endJunction** [comment]
@wenqing  I have extracted one commit for early merge in #603 with most line changes in there. The other things from PETSc and NeumannBC I shall merge later.

File: AssemblerLib/LocalToGlobalIndexMap.h [hunk 1]
```diff
@@ -110,6 +137,34 @@ class LocalToGlobalIndexMap
     /// For non-parallel implementations the columns are equal to the rows.
     /// \todo This is to be overriden by any parallel implementation.
     std::vector<LineIndex> const& _columns = _rows;
+
+#ifndef NDEBUG
+    friend std::ostream& operator<<(std::ostream& os, LocalToGlobalIndexMap& map)
```
Comment (TomFischer) [inline AssemblerLib/LocalToGlobalIndexMap.h]: Is it possible to make the second argument (```map```) a const reference?

File: AssemblerLib/MeshComponentMap.cpp [hunk 2]
```diff
@@ -48,6 +48,42 @@ MeshComponentMap::MeshComponentMap(
...
+
+    std::size_t comp_id = 0;
+    for (auto c : components)
+    {
+        if (c == nullptr)   // Empty component
+        {
+            comp_id++;
+            continue;
+        }
+        for (unsigned mesh_subset_index = 0; mesh_subset_index < c->size(); mesh_subset_index++)
```
Comment (TomFischer) [inline AssemblerLib/MeshComponentMap.cpp]: Why not ```std::size_t``` instead of ```unsigned``` here?

File: MeshLib/MeshSubset.h [hunk 3]
```diff
@@ -91,6 +103,25 @@ class MeshSubset
         return _msh.getElements().cend();
     }
 
+    MeshSubset*
+    setIntersectionByNodes(std::vector<Node*> const& nodes) const
```
Comment (TomFischer) [inline MeshLib/MeshSubset.h]: Is it possible to assume that ```nodes``` are sorted by their addresses?

File: Tests/AssemblerLib/TestMeshComponentMap.cpp [hunk 4]
```diff
@@ -136,3 +136,72 @@ TEST_F(AssemblerLibMeshComponentMapTest, OutOfRangeAccess)
     ASSERT_EQ(MeshComponentMap::nop, cmap->getGlobalIndex(
         Location(mesh->getID(), MeshItemType::Node, 0), 10));
 }
+
+TEST_F(AssemblerLibMeshComponentMapTest, SubsetOfNodesByComponent)
+{
+    cmap = new MeshComponentMap(components,
+        AssemblerLib::ComponentOrder::BY_COMPONENT);
+
+    //std::cout << "# database \n" << *cmap << std::endl;
```
Comment (TomFischer) [inline Tests/AssemblerLib/TestMeshComponentMap.cpp]: Is this commented code important? Same for line 161, 179, 195.

File: Tests/AssemblerLib/LocalToGlobalIndexMap.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    // components.
+    ASSERT_EQ(mesh->getNNodes() * components.size(), dof_map->dofSize());
+}
+
+TEST_F(AssemblerLibLocalToGlobalIndexMapTest, SubsetByComponent)
+{
+    dof_map = new AssemblerLib::LocalToGlobalIndexMap(components,
+        AssemblerLib::ComponentOrder::BY_COMPONENT);
+
+    std::cout << "# dof table\n" << *dof_map << std::endl;
```
Comment (TomFischer) [inline Tests/AssemblerLib/LocalToGlobalIndexMap.cpp]: Is this output for debug purposes?

File: Tests/AssemblerLib/LocalToGlobalIndexMap.cpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    // Select some elements from the full mesh.
+    std::array<std::size_t, 3> const ids = {{ 0, 5, 9 }};
+    std::vector<MeshLib::Element*> some_elements;
+    for (std::size_t id : ids)
+        some_elements.push_back(const_cast<MeshLib::Element*>(mesh->getElement(id)));
+
+    AssemblerLib::LocalToGlobalIndexMap* dof_map_subset =
+        dof_map->deriveBoundaryConstrainedMap(components, some_elements);
```
Comment (TomFischer) [inline Tests/AssemblerLib/LocalToGlobalIndexMap.cpp]: (1) What are you testing here?
(2) Is ```dof_map_subset``` deleted somewhere?

**Comment by TomFischer** [comment]
Nice approach to reuse existing things for Neumann BCs. The code looks good to me. Imho we can merge the PR apart from small comments.

File: AssemblerLib/MeshComponentMap.h [hunk 7]
```diff
@@ -129,6 +134,17 @@ class MeshComponentMap
 #endif  // NDEBUG
 
 private:
+    /// Private constructor used by internally created mesh component maps.
+    MeshComponentMap(detail::ComponentGlobalIndexDict& dict)
+        : _dict(dict)
+    { }
+
+    /// Looks up of a line already stored in the dictionary.
```
Comment (bilke) [inline AssemblerLib/MeshComponentMap.h]: Typo: Looks up **if** a line **is** already stored...

**Comment by bilke** [comment]
:+1:

File: MeshLib/MeshSubset.h (hunk 3 above)
Comment (endJunction) [inline MeshLib/MeshSubset.h]: @TomFischer No, the assumption can not be made. One could sort the input `nodes` internally, though. But for currently planned usage it will not change anything significant.
To use `std::set_intersection` both ranges must be sorted, which is pulling some more changes.

**Comment by endJunction** [comment]
I have finished the test for the dof table subset. It required another commit 
https://github.com/endJunction/ogs/commit/0322e4f2d56d9b4015d1c64ae5456b4bd822f6c4
providing selection of nodes used by elements.
So far all issues in the PR above were corrected.

File: AssemblerLib/LocalToGlobalIndexMap.h [hunk 8]
```diff
@@ -110,6 +137,34 @@ class LocalToGlobalIndexMap
     /// For non-parallel implementations the columns are equal to the rows.
     /// \todo This is to be overriden by any parallel implementation.
     std::vector<LineIndex> const& _columns = _rows;
+
+#ifndef NDEBUG
+    friend std::ostream& operator<<(std::ostream& os, LocalToGlobalIndexMap const& map)
```
Comment (norihiro-w) [inline AssemblerLib/LocalToGlobalIndexMap.h]: would you please move this definition to *.cpp?

File: AssemblerLib/LocalToGlobalIndexMap.h [hunk 9]
```diff
@@ -53,6 +57,19 @@ class LocalToGlobalIndexMap
         AssemblerLib::ComponentOrder const order =
             AssemblerLib::ComponentOrder::BY_COMPONENT);
 
+    // The elements are not necessary those used in the mesh_subsets.
```
Comment (norihiro-w) [inline AssemblerLib/LocalToGlobalIndexMap.h]: ///

Comment (norihiro-w) [inline AssemblerLib/LocalToGlobalIndexMap.h]: can you please document parameters?

File: AssemblerLib/MeshComponentMap.cpp [hunk 10]
```diff
@@ -48,6 +48,39 @@ MeshComponentMap::MeshComponentMap(
...
+                subset_dict.insert(getLine(Location(mesh_id,
+                    MeshLib::MeshItemType::Node, mesh_subset.getNodeID(j)), comp_id));
+            for (std::size_t j=0; j<mesh_subset.getNElements(); j++)
+                subset_dict.insert(getLine(Location(mesh_id,
+                    MeshLib::MeshItemType::Cell, mesh_subset.getElementID(j)), comp_id));
+        }
+        comp_id++;
+    }
+
+    return MeshComponentMap(subset_dict);
```
Comment (norihiro-w) [inline AssemblerLib/MeshComponentMap.cpp]: why not new MeshComponentMap here? Otherwise, the object is copied

File: AssemblerLib/MeshComponentMap.h [hunk 11]
```diff
@@ -122,13 +127,25 @@ class MeshComponentMap
...
+    /// Private constructor used by internally created mesh component maps.
+    MeshComponentMap(detail::ComponentGlobalIndexDict& dict)
+        : _dict(dict)
+    { }
+
+    /// Looks up if a line is already stored in the dictionary.
+    /// \attention The line for the location l and component c must exist,
+    /// the behaviour is undefined otherwise.
+    /// \return a copy of the line.
+    detail::Line getLine(Location const& l, std::size_t const c) const;
```
Comment (norihiro-w) [inline AssemblerLib/MeshComponentMap.h]: what is `c`? is it component id?

File: MeshLib/ComponentSelector.h [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+#define MESHLIB_COMPONENTSELECTOR_H_
+
+namespace MeshLib
+{
+
+
+/// Create a vector of unique nodes used by given elements.
+inline
+std::vector<MeshLib::Node*>
+selectNodes(std::vector<MeshLib::Element*> const& elements)
```
Comment (norihiro-w) [inline MeshLib/ComponentSelector.h]: how about moving this function to MeshSearcher.h?

File: MeshLib/MeshSubset.h [hunk 13]
```diff
@@ -81,6 +81,18 @@ class MeshSubset
         return (_eles==nullptr) ? 0 : _eles->size();
     }
 
+    /// Returns the global element id Element::getID() of i-th element in the
+    /// mesh subset.
+    /// Throws std::out_of_range exception if there are no nodes available.
+    std::size_t getElementID(std::size_t const i) const
+    {
+        if (!_eles)
```
Comment (norihiro-w) [inline MeshLib/MeshSubset.h]: isn't it enough to check this in debug mode?

File: MeshLib/MeshSubset.h (hunk 3 above)
Comment (norihiro-w) [inline MeshLib/MeshSubset.h]: please add documentation

Comment (norihiro-w) [inline MeshLib/MeshSubset.h]: the function name is confusing because it does not set anything to this object. I suggest something like `getIntersectionByNodes()`

File: AssemblerLib/LocalToGlobalIndexMap.cpp [hunk 14]
```diff
@@ -35,6 +35,26 @@ LocalToGlobalIndexMap::LocalToGlobalIndexMap(
...
+{
+    // For all MeshSubsets and each of their MeshSubset's and each element
+    // of that MeshSubset save a line of global indices.
+    for (MeshLib::MeshSubsets const* const mss : _mesh_subsets)
+    {
+        for (MeshLib::MeshSubset const* const ms : *mss)
+        {
+            std::size_t const mesh_id = ms->getMeshID();
+
+            findGlobalIndices(elements.cbegin(), elements.cend(), mesh_id, order);
```
Comment (wenqing) [inline AssemblerLib/LocalToGlobalIndexMap.cpp]: I have to replace the third argument of this function with ``MeshLib::MeshSubset const* const mesh_subset`` instead of  `` std::size_t const mesh_id``.

File: AssemblerLib/LocalToGlobalIndexMap.h [hunk 15]
```diff
@@ -65,6 +82,16 @@ class LocalToGlobalIndexMap
...
+    /// Private constructor used by internally created local-to-global index
+    /// maps. The mesh_component_map is passed as argument instead of being
+    /// created by the constructor.
+    explicit LocalToGlobalIndexMap(
+        std::vector<MeshLib::MeshSubsets*> const& mesh_subsets,
+        std::vector<MeshLib::Element*> const& elements,
+        AssemblerLib::MeshComponentMap const& mesh_component_map,
+        AssemblerLib::ComponentOrder const order);
+
     template <typename ElementIterator>
```
Comment (wenqing) [inline AssemblerLib/LocalToGlobalIndexMap.h]: I'd like to move this function definition to the place after the class declaration.

File: AssemblerLib/LocalToGlobalIndexMap.cpp (hunk 14 above)
Comment (wenqing) [inline AssemblerLib/LocalToGlobalIndexMap.cpp]: c.f. https://github.com/wenqing/ogs/blob/pFEM4/AssemblerLib/LocalToGlobalIndexMap.h

File: AssemblerLib/LocalToGlobalIndexMap.h (hunk 15 above)
Comment (TomFischer) [inline AssemblerLib/LocalToGlobalIndexMap.h]: > I'd like to move this function definition to the place after the class declaration.

@wenging What is the motivation for the request?

File: AssemblerLib/MeshComponentMap.cpp (hunk 10 above)
Comment (TomFischer) [inline AssemblerLib/MeshComponentMap.cpp]: I assume the compiler will omit the copy using return value optimization.

File: AssemblerLib/LocalToGlobalIndexMap.h (hunk 15 above)
Comment (wenqing) [inline AssemblerLib/LocalToGlobalIndexMap.h]: @TomFischer  Just think it is a little bit big function inside the class deflations. Besides I have to added some lines inside it ( #580). Looks like #580 will be merged after this PR. Otherwise, I would like you to take a look at #580.

File: AssemblerLib/MeshComponentMap.cpp (hunk 10 above)
Comment (endJunction) [inline AssemblerLib/MeshComponentMap.cpp]: At this point, as Tom added, there is no copy because of the RVO, which will be done in the debugging mode too: On lower level of compilation the address of the object from the caller side is passed (like hidden parameter), and then used for object construction here.

But while looking into this potential copy here, I could not avoid the call to the copy ctor. And that call is not here, but in initialization list of `LocalToGlobalIndexMap` [line 43 in the cpp file](https://github.com/ufz/ogs/pull/599/files#diff-3d365cd12eafd459267739cd14224f74R43).
I shall try to get rid of it.

File: AssemblerLib/LocalToGlobalIndexMap.h (hunk 15 above)
Comment (endJunction) [inline AssemblerLib/LocalToGlobalIndexMap.h]: I agree, the function inside the class definition is not small. We shall move it or split it when @wenqing changes are going to be merged.
